### PR TITLE
Implemented Queue Delete Message

### PR DESCRIPTION
**Pull request by Alexei-B**
PR to address #21 

To complete the message lifecycle for Azure Storage Queue we need the ability to delete messages.
To enable this, I have made the `Message` struct public, and used it as the required parameter to the new `DeleteMessageBuilder` request's method `with_message`. This sets the `messageid` URL path parameter, as well as the `popreceipt` query parameter.
I have added an example to show how the API can be used.

### CODE REVIEW

File: sdk/storage/src/queue/mod.rs
Comment (MindFlavor) [inline sdk/storage/src/queue/mod.rs:92]: I like this idea. 
I would prefer if we use a *smaller* struct/trait for this (`Message` has a lot of fields that are useless in this context). 

Something like:

```rust
pub trait PopReceipt {
  fn message_id(&self) -> &str;  
  fn pop_receipt(&self) -> &str;
}
```

And we could implement `PopReceipt` for `Message`.

The API would become:

```rust 
pub trait PopRecepitSupport<'b> {
    type O;
    fn with_pop_receipt(self, message: &'b dyn PopReceipt) -> Self::O;
}

pub trait PopRecepitRequired {
    fn pop_receipt(&self) -> &dyn PopReceipt;
}
```

This way it could be used with `Message` as before but also, potentially, by a struct that has lesser fields. 

What do you think?

**Review by MindFlavor — APPROVED** [review]
Nice work, thank you!

I just have one suggestion to make the API cleaner but it's optional (we can add it a later time). Just tell me if you want me to merge this as it is and I'll do it.

File: sdk/storage/src/queue/mod.rs
Comment (Alexei-B) [reply sdk/storage/src/queue/mod.rs:92]: That seems sensible, I can update this PR with that.

Comment (Alexei-B) [reply sdk/storage/src/queue/mod.rs]: I've updated the PR with that change. Let me know what you think!
Also, while testing this I found a bug, where pop_receipt can contain a '+' character. This gets interpreted as a space in the URL, meaning that the pop receipt became invalid. I have fixed this by adding percent encoding for that field.

Comment (Alexei-B) [reply sdk/storage/src/queue/mod.rs]: The part of this PR I think that I'm least sure about right now is using `Box<dyn PopReceipt>`, which I chose simply because I managed to convince the compiler that, that was correct. Still pretty new to rust so if that isn't appropriate then please let me know!

Comment (MindFlavor) [reply sdk/storage/src/queue/mod.rs]: It's ok! Whenever to own or to borrow is a subject of endless debate 😉, we'll address it if the time comes.

File: sdk/storage/src/queue/requests/delete_message_builder.rs
Comment (MindFlavor) [inline sdk/storage/src/queue/requests/delete_message_builder.rs:135]: It seems it would be possibile to call `execute` without having set the mandatory `pop_receipt`. This will cause a runtime panic (the `unwrap` on the getter). 

We can fix this after this one is merged.

Comment (Alexei-B) [reply sdk/storage/src/queue/requests/delete_message_builder.rs:135]: Ahh right, that'll be all the `Yes` and `No` stuff in `PutMessageBuilder` right? I was wondering what that did.

**Comment by Alexei-B** [comment]
> Nice work, thank you!
> 
> I just have one suggestion to make the API cleaner but it's optional (we can add it a later time). Just tell me if you want me to merge this as it is and I'll do it.

I'm ready for this to be merged, happy to add the protection against calling execute without a pop receipt if you don't beat me to it.

File: sdk/storage/src/queue/requests/delete_message_builder.rs
Comment (MindFlavor) [reply sdk/storage/src/queue/requests/delete_message_builder.rs:135]: Yes, I follow this approach for compile-time enforced builders: [https://dev.to/mindflavor/rust-builder-pattern-with-types-3chf](https://dev.to/mindflavor/rust-builder-pattern-with-types-3chf).

This is nice for the user perspective but adds a lot of boilerplate code to the crate maintainers.

**Comment by MindFlavor** [comment]
> I'm ready for this to be merged, happy to add the protection against calling execute without a pop receipt if you don't beat me to it.

Ok as soon as the checks complete I'll merge it and open the issue for the compile-time checks. 
If you are up to it I'll leave you to implement it (it's your code after all, you deserve all the credit 👍!)

**Pull request merged**
